x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,562 @@
+        super.doStop();
+        this.delegatedRepository.stop();
+    }
+
+    @Override
+    protected void doClose() {
+        super.doClose();
+        this.delegatedRepository.close();
+    }
+
+    private static Supplier<Tuple<SecretKey, String>> createDEKGenerator() throws GeneralSecurityException {
+        // DEK and DEK Ids are generated randomly
+        final SecureRandom DEKSecureRandom = SecureRandom.getInstance(RAND_ALGO);
+        final SecureRandom DEKIdSecureRandom = SecureRandom.getInstance(RAND_ALGO);
+        final KeyGenerator dataEncryptionKeyGenerator = KeyGenerator.getInstance(DATA_ENCRYPTION_SCHEME.split("/")[0]);
+        dataEncryptionKeyGenerator.init(AESKeyUtils.KEY_LENGTH_IN_BYTES * Byte.SIZE, DEKSecureRandom);
+        return () -> new Tuple<>(dataEncryptionKeyGenerator.generateKey(), UUIDs.randomBase64UUID(DEKIdSecureRandom));
+    }
+
+    static class EncryptedBlobStore implements BlobStore {
+        private final BlobStore delegatedBlobStore;
+        private final BlobPath delegatedBasePath;
+        private final Function<String, Tuple<String, SecretKey>> getKEKforDEK;
+        private final CheckedSupplier<SingleUseDEK, IOException> singleUseDEKSupplier;
+        private final CheckedFunction<String, SecretKey, IOException> getDEKById;
+
+        EncryptedBlobStore(BlobStore delegatedBlobStore,
+                           BlobPath delegatedBasePath,
+                           char[] repositoryPassword,
+                           Supplier<Tuple<SecretKey, String>> DEKGenerator,
+                           Cache<String, SecretKey> DEKCache) {
+            this.delegatedBlobStore = delegatedBlobStore;
+            this.delegatedBasePath = delegatedBasePath;
+            this.getKEKforDEK = DEKId -> {
+                try {
+                    SecretKey KEK = AESKeyUtils.generatePasswordBasedKey(repositoryPassword,
+                            Base64.getUrlDecoder().decode(DEKId.getBytes(StandardCharsets.UTF_8)));
+                    String KEKId = AESKeyUtils.computeId(KEK);
+                    return new Tuple<>(KEKId, KEK);
+                } catch (GeneralSecurityException e) {
+                    throw new ElasticsearchException("Failure to generate KEK to wrap the DEK [" + DEKId + "]", e);
+                }
+            };
+            this.singleUseDEKSupplier = SingleUseDEK.createSingleUseDEKSupplier(() -> {
+                Tuple<SecretKey, String> newDEK = DEKGenerator.get();
+                // store newly generated DEK before making it available
+                storeDEK(newDEK.v2(), newDEK.v1());
+                return newDEK;
+            });
+            this.getDEKById = DEKId -> {
+                try {
+                    return DEKCache.computeIfAbsent(DEKId, ignored -> loadDEK(DEKId));
+                } catch (ExecutionException e) {
+                    // some exception types are to be expected
+                    if (e.getCause() instanceof IOException) {
```

Comment:
I used it because the `loadDEK` method is careful to throw `IOExceptions` when storage has problems and `RepositoryException` when there are encryption problems (wrong password or tampered metadata). I've learned that `IOException`s in `readBlob` can move the repository in a corrupted state, and if we wrap the `IOException` for DEK load (which is called before reading the actual blob) that'd amount to a change of behavior in those cases.

Moreover, I don't believe the `ExecutionException` wrapping adds useful information.